File: src/registry/schedule.rs
use k256::ecdsa::SigningKey;

use super::*;

// --- OFF-CHAIN scheduling (GitHub store via the proxy) ----------------------
//
// Scheduled jobs live OFF-CHAIN: the browser/CLI POST `/api/schedule` and the
// proxy keeps the job as a GitHub-store record (no escrow, no gas, no sponsor —
// the same off-chain model as the app store / telemetry). A `reminder` is a
// future web-push (zero $LH); an `agent` job runs the target each fire, billed
// per run from the OWNER's meter (same cost as an interactive message). The
// cron worker fires both stores. Auth = the personal-sign proxy token; the
// signer is the job OWNER (billing + push). The on-chain ScheduleFacet client
// (escrowed jobs, keeper drain) was PURGED 2026-07-30 — pre-1.0.0 reset, no
// jobs to drain; the facet itself dies at the reset genesis.

/// Create an off-chain scheduled job via `POST /api/schedule`. `kind` =
/// `"reminder"` | `"agent"`; `target` is the subdomain to run (agent only —
/// ignored/empty for a reminder). Returns the new job id. `now_secs` is the
/// caller's UNIX time, passed in so this is cross-target (native CLI =
/// `SystemTime`, browser = `js_sys::Date::now()`).
pub async fn create_offchain_job(
    signer: &SigningKey,
    now_secs: u64,
    kind: &str,
    target: &str,
    task: &str,
    interval_secs: u64,
    runs: u32,
) -> Result<String, String> {
    let token = proxy_auth_token(signer, now_secs, "schedule");
    let url = format!("{CREDIT_PROXY_URL}api/schedule");
    let mut body = serde_json::json!({
        "action": "create",
        "kind": kind,
        "task": task,
        "intervalSecs": interval_secs,
        "runs": runs,
    });
    if !target.is_empty() {
        body["target"] = serde_json::Value::String(target.to_string());
    }
    let resp = http_post_json_authed_returning(&url, &token, &body).await?;
    resp.get("id")
        .and_then(|v| v.as_str())
        .map(|s| s.to_string())
        .ok_or_else(|| format!("schedule: unexpected response {resp}"))
}

/// Cancel an off-chain scheduled job by id (owner-gated server-side: the proxy
/// only deletes a job whose stored owner matches the token's signer). `Ok(())`
/// on success; an unknown id / not-yours surfaces the proxy's error.
pub async fn cancel_offchain_job(signer: &SigningKey, now_secs: u64, id: &str) -> Result<(), String> {
    let token = proxy_auth_token(signer, now_secs, "schedule");
    let url = format!("{CREDIT_PROXY_URL}api/schedule");
    let body = serde_json::json!({ "action": "cancel", "id": id });
    http_post_json_authed_returning(&url, &token, &body).await.map(|_| ())
}

/// List the caller's off-chain scheduled jobs (the `jobs` array). Read-only.
pub async fn list_offchain_jobs(
    signer: &SigningKey,
    now_secs: u64,
) -> Result<Vec<serde_json::Value>, String> {
    let token = proxy_auth_token(signer, now_secs, "schedule");
    let url = format!("{CREDIT_PROXY_URL}api/schedule");
    let body = serde_json::json!({ "action": "list" });
    let resp = http_post_json_authed_returning(&url, &token, &body).await?;
    Ok(resp
        .get("jobs")
        .and_then(|v| v.as_array())
        .cloned()
        .unwrap_or_default())
}

File: proxy/api/schedule.ts
// /api/schedule — OFF-CHAIN scheduled-job WRITE path (create / cancel / list).
//
// The browser `schedule_task`/`cancel_task` tools and the CLI `schedule`/`goal`/
// `jobs`/`unschedule` POST here instead of submitting a sponsored Tempo tx to
// ScheduleFacet. Jobs are off-chain data (see `_jobstore.ts`): no gas, no
// escrow, no sponsor drain. A `reminder` job is just a future web-push (zero
// chain, zero $LH); an `agent` job runs the target agent each fire, billed per
// run from the OWNER's existing meter (same cost as an interactive message — no
// schedule tax). The cron worker (`scheduler.ts`) fires the due set.
//
// Auth = the SAME personal-sign token as gemini.ts/publish.ts/telemetry.ts
// (`address:timestamp:signature` in `x-goog-api-key`, 300s freshness). The
// authed address is the job OWNER — the billing + push identity. cancel/list
// only ever touch the caller's OWN jobs (owner match).

import { verifyAuthToken, ethCall, selector, isAllowedOrigin } from './_auth';
import { bytesToHex } from '@noble/hashes/utils';
import {
  createJob,
  findById,
  listByOwner,
  countByOwner,
  deleteJob,
  jobStoreConfigured,
  MAX_TASK_BYTES,
  MIN_INTERVAL_SECS,
  MAX_RUNS,
  MAX_JOBS_PER_OWNER,
  type JobKind,
  type OffchainJob,
} from './_jobstore';

export const config = { runtime: 'edge' };

// --- CORS (same policy as publish.ts / telemetry.ts) -------------------------
function corsHeaders(origin: string | null): Record<string, string> {
  const h: Record<string, string> = {
    'Access-Control-Allow-Methods': 'POST, OPTIONS',
    'Access-Control-Allow-Headers': 'content-type, x-goog-api-key, x-api-key',
    Vary: 'Origin',
  };
  if (origin && isAllowedOrigin(origin)) h['Access-Control-Allow-Origin'] = origin;
  return h;
}
function json(body: unknown, status: number, origin: string | null): Response {
  return new Response(JSON.stringify(body), {
    status,
    headers: { 'content-type': 'application/json', ...corsHeaders(origin) },
  });
}

/** ABI-encode a single `string` arg (offset 0x20 | length | utf8 padded) —
 * mirrors publish.ts (idOfName takes one string). */
function encodeStringArg(value: string): string {
  const bytes = new TextEncoder().encode(value);
  const len = bytes.length;
  const padded = Math.ceil(len / 32) * 32;
  const buf = new Uint8Array(32 + 32 + padded);
  buf[31] = 0x20;
  let x = len;
  for (let i = 63; i >= 32 && x > 0; i--) {
    buf[i] = x & 0xff;
    x = Math.floor(x / 256);
  }
  buf.set(bytes, 64);
  return bytesToHex(buf);
}

/** `idOfName(string) -> uint256`. 0n = unregistered. */
async function idOfName(name: string): Promise<bigint> {
  const res = await ethCall('0x' + selector('idOfName(string)') + encodeStringArg(name));
  try {
    return BigInt(res);
  } catch {
    return 0n;
  }
}

export default async function handler(req: Request): Promise<Response> {
  const origin = req.headers.get('origin');
  if (req.method === 'OPTIONS') return new Response(null, { status: 204, headers: corsHeaders(origin) });
  if (req.method !== 'POST') return json({ error: 'POST only' }, 405, origin);
  if (!jobStoreConfigured()) return json({ error: 'scheduler store not configured (no GitHub token)' }, 503, origin);

  // Auth — personal-sign token (address:ts:sig), 300s freshness. The authed
  // address is the job owner (billing + push identity).
  const token = req.headers.get('x-goog-api-key') ?? req.headers.get('x-api-key') ?? '';
  const now = Math.floor(Date.now() / 1000);
  const auth = verifyAuthToken(token, now);
  if (!auth.ok) return json({ error: 'auth: ' + auth.error }, auth.status, origin);
  const owner = auth.address.toLowerCase();

  let payload: Record<string, unknown>;
  try {
    payload = await req.json();
  } catch {
    return json({ error: 'bad json' }, 400, origin);
  }

  const action = String(payload.action ?? '').trim();

  // ---- cancel: delete the caller's OWN job by id --------------------------
  if (action === 'cancel') {
    const id = String(payload.id ?? '').trim();
    if (!id) return json({ error: 'cancel requires "id"' }, 400, origin);
    let found: { job: OffchainJob; path: string } | null;
    try {
      found = await findById(id);
    } catch (e) {
      return json({ error: 'store: ' + (e as Error).message }, 502, origin);
    }
    if (!found) return json({ error: `no job ${id}` }, 404, origin);
    if (found.job.owner.toLowerCase() !== owner) {
      return json({ error: 'not your job' }, 403, origin);
    }
    try {
      await deleteJob(found.path, `cancel job ${id} (owner ${owner.slice(0, 10)})`);
    } catch (e) {
      return json({ error: 'store: ' + (e as Error).message }, 502, origin);
    }
    return json({ cancelled: true, id }, 200, origin);
  }

  // ---- list: the caller's OWN jobs ----------------------------------------
  if (action === 'list') {
    try {
      const jobs = await listByOwner(owner);
      return json({ jobs }, 200, origin);
    } catch (e) {
      return json({ error: 'store: ' + (e as Error).message }, 502, origin);
    }
  }

  // ---- create -------------------------------------------------------------
  if (action !== 'create') {
    return json({ error: 'action must be "create", "cancel", or "list"' }, 400, origin);
  }

  const kind: JobKind = payload.kind === 'agent' ? 'agent' : 'reminder';
  const task = String(payload.task ?? '').trim();
  if (!task) return json({ error: 'create requires a non-empty "task"' }, 400, origin);
  if (new TextEncoder().encode(task).length > MAX_TASK_BYTES) {
    return json({ error: `task too large (> ${MAX_TASK_BYTES} bytes)` }, 413, origin);
  }

  const intervalSecs = Math.trunc(Number(payload.intervalSecs ?? payload.interval_secs ?? 0));
  if (!Number.isFinite(intervalSecs) || intervalSecs < MIN_INTERVAL_SECS) {
    return json({ error: `intervalSecs must be >= ${MIN_INTERVAL_SECS}` }, 400, origin);
  }
  const runs = Math.trunc(Number(payload.runs ?? 1));
  if (!Number.isFinite(runs) || runs < 1 || runs > MAX_RUNS) {
    return json({ error: `runs must be 1..${MAX_RUNS}` }, 400, origin);
  }

  // For an AGENT job, resolve the target name → tokenId (persona + push slot).
  // For a REMINDER, no target/model — the push goes to the OWNER's subscription
  // (the cron resolves mainOf(owner)).
  let target = '';
  let targetId = '0';
  if (kind === 'agent') {
    target = String(payload.target ?? '').trim().toLowerCase();
    if (!/^[a-z0-9-]{1,63}$/.test(target)) {
      return json({ error: 'agent job requires a valid "target" subdomain' }, 400, origin);
    }
    let id: bigint;
    try {
      id = await idOfName(target);
    } catch (e) {
      return json({ error: 'rpc: ' + (e as Error).message }, 502, origin);
    }
    if (id === 0n) return json({ error: `"${target}" is not a registered agent` }, 404, origin);
    targetId = id.toString();
  }

  // Per-owner active-job cap — bounds the shared store so one keypair can't flood
  // it with free reminders and crowd the cron's due scan (finding: cross-tenant
  // DoS on a free creation primitive).
  let owned: number;
  try {
    owned = await countByOwner(owner);
  } catch (e) {
    return json({ error: 'store: ' + (e as Error).message }, 502, origin);
  }
  if (owned >= MAX_JOBS_PER_OWNER) {
    return json(
      { error: `you have ${owned} scheduled jobs (max ${MAX_JOBS_PER_OWNER}) — cancel some first` },
      429,
      origin,
    );
  }

  const job: OffchainJob = {
    id: crypto.randomUUID(),
    owner,
    kind,
    target,
    targetId,
    task,
    intervalSecs,
    runsLeft: runs,
    nextRun: now + intervalSecs,
    createdAt: now,
  };

  try {
    await createJob(job);
  } catch (e) {
    return json({ error: 'store: ' + (e as Error).message }, 502, origin);
  }

  return json(
    { scheduled: true, id: job.id, kind, target, intervalSecs, runs, nextRun: job.nextRun },
    200,
    origin,
  );
}
